### ISSUE TITLE
otlphttp exporter drops data because memory limiter (of another instance) is considered as non recoverable error

### ISSUE THREAD

**Issue by sumo-drosiek** [open-telemetry/opentelemetry-collector#8132]
**Describe the bug**

Data are being dropped by exporter, because receiver is throttled by memory limiter. This is incorrect in my opinion as receiver may be behind load balancer, so after retry packet may be forwarded to healthy instance. Also memory usage is not constant, and may not be subject of limitation after retry

**Steps to reproduce**

Not sure how to reproduce, we observe it in our environemnt when receiver instance is memory pressured

**What did you expect to see?**

I expect that receiver will try to resend data after some time, or put it back to sending queue

**What did you see instead?**

I see that data is dropped

**What version did you use?**

`v0.81.0`

**What config did you use?**

```
exporters:                                                                                                                                                                                                                                                                                                                                                                                                                            
  otlphttp:                                                                                                                                                                                                                                                                                                                                                                                                                           
    endpoint: ...                                                                                                                                                                                                                                                                                                                                                       
    retry_on_failure:                                                                                                                                                                                                                                                                                                                                                                                                                 
      max_elapsed_time: 600s                                                                                                                                                                                                                                                                                                                                                                                                          
    sending_queue:                                                                                                                                                                                                                                                                                                                                                                                                                    
      queue_size: 1000                                                                                                                                                                                                                                                                                                                                                                                                                
      storage: file_storage
```

**Environment**
N/A

**Additional context**
N/A

**Comment by neetle** [comment]
We're also seeing this issue with exporting traces via the GRPC version of this client when a memorylimiter is in play. 

Looking at the error logs, we're seeing:
```
error	exporterhelper/queued_retry.go:391	Exporting failed. The error is not retryable. Dropping data.	{"kind": "exporter", "data_type": "traces", "name": "otlp", "error": "Permanent error: rpc error: code = Unknown desc = data refused due to high memory usage", "dropped_items": n}
```

From my interpretation of the code & the logs, given an exporter in service A and a receiver & processor in service B:
1) The exporter in service A sends a request to the receiver in service B.
2) The receiver in service B passes the traces to the processor 
3) The processor in service B returns an error. This is propagated back up the pipeline to the receiver.
4) The receiver in service B gets the error. It runs it through https://github.com/open-telemetry/opentelemetry-collector/blob/3089ea878eef022913a053600f3599b3fddffb95/receiver/otlpreceiver/otlphttp.go#L157 or similar in the GRPC version. 
5) The error is mapped to a status code that is considered "unretryable" and returned to the exporter in service A.
6) Service A sees some form of "permanent" error and gives up. 

If the errors were marked as `Retryable() bool`/`Reason() ReasonEnum`  or similar, we could migrate the pipeline piecemeal without BC-breaking the existing `error == Permanent` behaviour.

**Comment by kristinnSnow** [comment]
From the memory limiter documentation we can see that this should be a retryable error and the if it doesn't try to retry, then the component is considered incorrectly implemented.

> In memory limited mode the error returned by ConsumeLogs/Trace/Metrics function is a non-permanent error. When receivers see this error they are expected to retry sending the same data. The receivers may also apply a backpressure to their data sources in order to slow down the inflow of data into the Collector and allow the memory usage to go below the limits.
>
> > Warning: if the component preceding the memory limiter in the pipeline does not correctly retry and send the data again after ConsumeLogs/Trace/Metrics functions return then that data will be permanently lost. We consider such components incorrectly implemented.

**Comment by neetle** [comment]
Do we have a way to mark errors within the pipeline as retry-able currently? Either on the error that we return, or in some form of config in the receiver?

**Comment by VijayakumarGn** [comment]
> Getting same error
> 
> Signoz [0.20.0 ](https://charts.signoz.io) from helm chart repository
> 
> ```
> otelCollector:
>   name: "otel-collector"
>   image:
>     registry: docker.io
>     repository: signoz/signoz-otel-collector
>     tag: 0.79.4
> ```
> 
> Running alone in a cluster with 21Go RAM + 6 CPU ...
> 
> ```
> 2023-07-26T13:45:58.924Z        info    otlpreceiver@v0.79.0/otlp.go:83 Starting GRPC server    {"kind": "receiver", "name": "otlp", "data_type": "traces", "endpoint": "0.0.0.0:4317"}
> 2023-07-26T13:45:58.924Z        warn    internal/warning.go:40  Using the 0.0.0.0 address exposes this server to every network interface, which may facilitate Denial of Service attacks        {"kind": "receiver", "name": "otlp", "data_type": "traces", "documentation": "https://github.com/open-telemetry/opentelemetry-collector/blob/main/docs/security-best-practices.md#safeguards-against-denial-of-service-attacks"}
> 2023-07-26T13:45:58.924Z        info    otlpreceiver@v0.79.0/otlp.go:101        Starting HTTP server    {"kind": "receiver", "name": "otlp", "data_type": "traces", "endpoint": "0.0.0.0:4318"}
> 2023-07-26T13:45:58.924Z        warn    internal/warning.go:40  Using the 0.0.0.0 address exposes this server to every network interface, which may facilitate Denial of Service attacks        {"kind": "receiver", "name": "jaeger", "data_type": "traces", "documentation": "https://github.com/open-telemetry/opentelemetry-collector/blob/main/docs/security-best-practices.md#safeguards-against-denial-of-service-attacks"}
> 2023-07-26T13:45:58.924Z        warn    internal/warning.go:40  Using the 0.0.0.0 address exposes this server to every network interface, which may facilitate Denial of Service attacks        {"kind": "receiver", "name": "jaeger", "data_type": "traces", "documentation": "https://github.com/open-telemetry/opentelemetry-collector/blob/main/docs/security-best-practices.md#safeguards-against-denial-of-service-attacks"}
> 2023-07-26T13:45:58.925Z        info    internal/resourcedetection.go:125       began detecting resource information    {"kind": "processor", "name": "resourcedetection", "pipeline": "metrics/internal"}
> 2023-07-26T13:45:58.926Z        info    internal/resourcedetection.go:139       detected resource information   {"kind": "processor", "name": "resourcedetection", "pipeline": "metrics/internal", "resource": {"host.id":"6981a73c-6bb5-43fd-b440-d942ecb09719","host.name":"signoz-otel-collector-869fc584f9-qxf8g","k8s.cluster.name":"","k8s.pod.ip":"10.2.1.22","k8s.pod.uid":"d9631bc8-b79f-4fd1-a0ec-24673e09f273","os.type":"linux","signoz.component":"otel-collector"}}
> 2023-07-26T13:45:58.926Z        info    otlpreceiver@v0.79.0/otlp.go:83 Starting GRPC server    {"kind": "receiver", "name": "otlp/spanmetrics", "data_type": "metrics", "endpoint": "localhost:12345"}
> 2023-07-26T13:45:58.926Z        info    healthcheck/handler.go:129      Health Check state change       {"kind": "extension", "name": "health_check", "status": "ready"}
> 2023-07-26T13:45:58.926Z        info    service/service.go:148  Everything is ready. Begin running and processing data.
> 2023-07-26T14:26:04.105Z        error   exporterhelper/queued_retry.go:391      Exporting failed. The error is not retryable. Dropping data.    {"kind": "exporter", "data_type": "metrics", "name": "clickhousemetricswrite", "error": "Permanent error: unsupported metric type", "dropped_items": 1}
> go.opentelemetry.io/collector/exporter/exporterhelper.(*retrySender).send
>         /home/runner/go/pkg/mod/go.opentelemetry.io/collector/exporter@v0.79.0/exporterhelper/queued_retry.go:391
> go.opentelemetry.io/collector/exporter/exporterhelper.(*metricsSenderWithObservability).send
>         /home/runner/go/pkg/mod/go.opentelemetry.io/collector/exporter@v0.79.0/exporterhelper/metrics.go:125
> go.opentelemetry.io/collector/exporter/exporterhelper.(*queuedRetrySender).start.func1
>         /home/runner/go/pkg/mod/go.opentelemetry.io/collector/exporter@v0.79.0/exporterhelper/queued_retry.go:195
> go.opentelemetry.io/collector/exporter/exporterhelper/internal.(*boundedMemoryQueue).StartConsumers.func1
>         /home/runner/go/pkg/mod/go.opentelemetry.io/collector/exporter@v0.79.0/exporterhelper/internal/bounded_memory_queue.go:47
> 2023-07-26T14:27:04.149Z        error   exporterhelper/queued_retry.go:391      Exporting failed. The error is not retryable. Dropping data.    {"kind": "exporter", "data_type": "metrics", "name": "clickhousemetricswrite", "error": "Permanent error: unsupported metric type", "dropped_items": 1}
> go.opentelemetry.io/collector/exporter/exporterhelper.(*retrySender).send
>         /home/runner/go/pkg/mod/go.opentelemetry.io/collector/exporter@v0.79.0/exporterhelper/queued_retry.go:391
> go.opentelemetry.io/collector/exporter/exporterhelper.(*metricsSenderWithObservability).send
>         /home/runner/go/pkg/mod/go.opentelemetry.io/collector/exporter@v0.79.0/exporterhelper/metrics.go:125
> go.opentelemetry.io/collector/exporter/exporterhelper.(*queuedRetrySender).start.func1
>         /home/runner/go/pkg/mod/go.opentelemetry.io/collector/exporter@v0.79.0/exporterhelper/queued_retry.go:195
> go.opentelemetry.io/collector/exporter/exporterhelper/internal.(*boundedMemoryQueue).StartConsumers.func1
>         /home/runner/go/pkg/mod/go.opentelemetry.io/collector/exporter@v0.79.0/exporterhelper/internal/bounded_memory_queue.go:47
> ```
> 
> Traces, Metrics and Logs sent with protocol `otlp_proto_http`, handle on port 4318.
I am hit with the same problem, Did you find the solution

**Comment by veyselsahin** [comment]
Any improvement? I have the same issue.

**Comment by rupeshnemade** [comment]
Any update ? We are losing important data whenever collector receives this error & this 'permanent' error is not retryable which makes things worse.

**Comment by SaundersR1999** [comment]
Is there a way to add a config flag that disables the permanent error determination and instead gives every error the full retry window?

In my team's setup with Tempo we are experiencing dropped traces when the Tempo Ingester shuts down. It isn't permanent and Tempo is then unable to receive the trace batches when it comes back up. We want to be protected from this and wouldn't care if we are retrying other actual permanent errors for the full duration of a retry window. I'm sure others aware of this bug may feel similarly about it.

**Comment by TylerHelmuth** [comment]
I believe this issue will be fixed by https://github.com/open-telemetry/opentelemetry-collector/pull/9357

**Comment by SaundersR1999** [comment]
> I believe this issue will be fixed by #9357

That would be incredible. Will test it out once it's merged

**Comment by diegoamaralpp** [comment]
Hi @SaundersR1999. Based on your last response, did you already made the test since the PR was merged? The problem for traces scenario persists to you? 

Thanks
